File: scala-hoots/src/main/java/io/hoots/match/Matcher.java
package io.hoots.match;

import io.hoots.fingerprint.domain.Point;
import io.hoots.fingerprint.domain.Hash;
import io.hoots.fingerprint.domain.HashMatch;
import io.hoots.fingerprint.domain.FingerPrint;
import io.hoots.match.domain.Matches;

import java.util.*;

/**
 * Created by rwadowski on 07.04.17.
 */
public class Matcher {

    private final Map<Hash, List<Point>> hashes;

    public Matcher() {
        this(new HashMap<>());
    }

    public Matcher(Map<Hash, List<Point>> hashes) {
        this.hashes = hashes;
    }

    public void update(FingerPrint print) {
        final Map<Hash, List<Point>> printHashes = print.getPrint();
        final Set<Hash> keys = printHashes.keySet();
        for(Hash k : keys) {
            if(hashes.containsKey(k)) {
                final List<Point> points = hashes.get(k);
                final List<Point> newPoints = printHashes.get(k);
                points.addAll(newPoints);
                hashes.put(k, points);
            } else {
                hashes.put(k, printHashes.get(k));
            }
        }
    }

    public Matches process(FingerPrint print) {
        final Map<Hash, List<Point>> sample = print.getPrint();
        final List<HashMatch> matches = new ArrayList<>();
        final Set<Hash> sampleHashes = sample.keySet();
        for(Hash sh : sampleHashes) {
            if(!hashes.containsKey(sh)) {
                continue;
            }
            final List<Point> samplePoints = sample.get(sh);
            final List<Point> points = hashes.get(sh);
            final HashMatch m = new HashMatch(sh, samplePoints, points);
            matches.add(m);
        }
        return new Matches(matches, print.getInputFormat());
    }
}

File: scala-hoots/src/main/scala/io/hoots/results/domain/Matcher.scala
package io.hoots.results.domain

import io.hoots.fingerprint.domain.{FingerPrint, Hash, HashMatch, Point}

/**
  * Created by rwadowski on 19.04.17.
  */
class Matcher(hashesMap: Map[Hash, Seq[Point]] = Map.empty) {

  private lazy val hashes: Set[Hash] = hashesMap.keys.toSet

//  def +(fp: FingerPrint): Matcher = {
//
//    val m = collection.mutable.Map(hashesMap.toSeq: _*)
//    for((hash, newPoints) <- fp.print) {
//      if(m.contains(hash)) {
//        val currentPoints = m(hash)
//        m += hash -> (currentPoints ++ newPoints)
//      } else {
//        m += hash -> newPoints
//      }
//    }
//    new Matcher(m.toMap)
//  }

  def +(fp: FingerPrint): Matcher = {
    var map = hashesMap
    for((hash, newPoints) <- fp.print) {
      if(map.contains(hash)) {
        val currentPoints = map(hash)
        map = map + (hash -> (currentPoints ++ newPoints))
      } else {
        map = map + (hash -> newPoints)
      }
    }
    new Matcher(map)
  }

  def process(fp: FingerPrint): Matches = {
    val sample = fp.print
    val matched = hashes.intersect(sample.keys.toSet)

    val matchList = matched.map{h =>
      HashMatch(h, sample(h), hashesMap(h))
    }
    Matches(matchList.toSeq, fp.audioFormat)
  }
}
